### PR TITLE
[Basic] Add a CacheableSequence.

### PR DESCRIPTION
**Pull request by ddunbar**
- This adapts an arbitrary sequence to one which can be consumed multiple
   times.

 - I'm open to suggestions on a better name.

### CODE REVIEW

**Comment by aciidb0mb3r** [comment]
`MemoizedSequence` could be an alternative but I am fine with either.

File: Sources/Basic/CacheableSequence.swift [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    
+    /// An iterator on the underlying sequence, until complete.
+    fileprivate var it: T.Iterator?
+    
+    public init(_ sequence: T) {
+        self.it = sequence.makeIterator()
+    }
+    
+    public func makeIterator() -> Iterator {
+        return CacheableSequenceIterator(self)
```
Comment (aciidb0mb3r) [inline Sources/Basic/CacheableSequence.swift]: Can we not use any iterator?

File: Tests/BasicTests/CacheableSequenceTests.swift [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+ Copyright (c) 2018 Apple Inc. and the Swift project authors
+ Licensed under Apache License v2.0 with Runtime Library Exception
+
+ See http://swift.org/LICENSE.txt for license information
+ See http://swift.org/CONTRIBUTORS.txt for Swift project authors
+*/
+
+import XCTest
+import Dispatch
```
Comment (aciidb0mb3r) [inline Tests/BasicTests/CacheableSequenceTests.swift]: This isn't needed.

File: Sources/Basic/CacheableSequence.swift (hunk 1 above)
Comment (ddunbar) [reply Sources/Basic/CacheableSequence.swift]: It is better not to, this has stronger type information... clients can erase it if need be. I think, maybe you had something else in mind?

**Comment by ddunbar** [comment]
@swift-ci please smoke test

File: Sources/Basic/CacheableSequence.swift (hunk 1 above)
Comment (aciidb0mb3r) [reply Sources/Basic/CacheableSequence.swift]: Not sure if iterator type is really needed. AnyIterator does retain the element type.

**Pull request merged**